src/datasets/arrow_dataset.py

```diff
@@ -1559,6 +1555,12 @@ def apply_function_on_filtered_inputs(inputs, indices, check_same_num_examples=F
             if not batched:
                 for i, example in enumerate(pbar):
                     example = apply_function_on_filtered_inputs(example, i, offset=offset)
+                    # If cached, break and return a cached dataset, otherwise prepare resources
+                    if i == 0 and update_data:
+                        if not is_cached:
+                            buf_writer, writer, tmp_file = init_buffer_and_writer()
```

Comment:
I agree it would be more natural, but we have to know the value of `update_data` (to calculate `cache_file_name`) to correctly initialize these resources and we know that value after processing the first example/batch, which means we have to enter the for loop.